### PR TITLE
New Streaming and Buffered Interfaces

### PR DESCRIPTION
**Pull request by diasdavid**
- Files
  - [x] Review and declare new function signatures
  - [x] add 
    - [x] review spec
    - [x] update tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] addReadableStream
    - [x] review spec
    - [x] update tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] addPullStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] cat 
    - [x] review spec
    - [x] update tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] catReadableStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] catPullStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] get 
    - [x] review spec
    - [x] update tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] getReadableStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] getPullStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] ls 
    - [x] review spec
    - [x] update tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] lsReadableStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [x] lsPullStream
    - [x] review spec
    - [x] create tests
    - [x] impl js-ipfs-api
    - [x] impl js-ipfs
  - [ ] ~~merge addFromFS to add~~ will do in another PR
  - [ ] ~~merge addFromUrl to add~~ will do in another PR

### CODE REVIEW

File: SPEC/FILES.md [hunk 1]
```diff
@@ -58,13 +60,64 @@ ipfs.files.add(files, function (err, files) {
...
 
-#### `createAddStream`
+#### `AddReadableStream`
 
 > Add files and data to IPFS using a transform stream.
 
 ##### `Go` **WIP**
 
-##### `JavaScript` - ipfs.files.createAddStream([options], [callback])
+##### `JavaScript` - ipfs.files.AddReadableStream([options], [callback])
```
Comment (dignifiedquire) [inline SPEC/FILES.md]: please don't capitalize method names in JavaScript, that makes for some very surprising apis.

Comment (diasdavid) [reply SPEC/FILES.md]: ups, typo

File: SPEC/FILES.md [hunk 2]
```diff
@@ -52,142 +57,342 @@ const files = [
...
+const validCID = 'QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF'
 
-ipfs.files.get(multihashStr, function (err, stream) {
-  stream.on('data', (file) => {
-    // write the file's path and contents to standard out
+ipfs.files.get(validCID, function (err, files) {
+  files.forEach((file) => {
     console.log(file.path)
-    file.content.pipe(process.stdout)
+    console.log(file.path.toString())
```
Comment (Beanow) [inline SPEC/FILES.md]: Logging the path twice?
What about the path and filesize so we touched the content buffer as well.

Comment (diasdavid) [reply SPEC/FILES.md]: Thank you for catching that!

File: SPEC/FILES.md [hunk 3]
```diff
@@ -9,24 +11,27 @@ files API
...
 ##### `JavaScript` - ipfs.files.add(data, [options], [callback])
 
-Where `data` may be
+Where `data` may be:
 
-- an array of objects, each of the form
+- a [`Buffer instance`][b]
+- a [`Readable Stream`][rs]
+- a [`Pull Stream`][ps]
+- a Path (caveat: will only work in Node.js)
```
[the remark this answers is not in the document]
Comment (diasdavid) [reply SPEC/FILES.md]: Sounds good!

**Comment by diasdavid** [comment]
# Spec and js-ipfs-api have been updated!

I've finished updating the spec and the implementation in js-ipfs-api. Next step is to implement the same interface in js-ipfs.

Please review as soon as possible to make sure that nothing gets missed. Overall I'm really happy with the simplicity of the new calls and I bet a lot more people will be able to use both modules without having to go through the learning of Node.js Streams and/or Pull Streams

File: SPEC/FILES.md [hunk 4]
```diff
@@ -391,6 +391,158 @@ pull(
...
+  type: 'file'
+}
+```
+
+**Example:**
+
+```JavaScript
+const validCID = 'QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF'
+
+const stream = ipfs.files.getReadableStream(validCID)
```
Comment (pgte) [inline SPEC/FILES.md]: Shouldn't this be `... ipfs.lsReadableStream(..` instead?

File: SPEC/FILES.md [hunk 5]
```diff
@@ -391,6 +391,158 @@ pull(
...
+  // write the file's path and contents to standard out
+  console.log(file.path)
+})
+```
+
+A great source of [examples][] can be found in the tests for this API.
+
+#### `lsPullStream`
+
+> Fetch a file or an entire directory tree from IPFS that is addressed by a valid IPFS Path. The files will be yielded as Readable Streams.
```
Comment (pgte) [inline SPEC/FILES.md]: s/Readable Streams/pull stream ?

File: SPEC/FILES.md [hunk 6]
```diff
@@ -391,6 +391,158 @@ pull(
...
+  type: 'file'
+}
+```
+
+**Example:**
+
+```JavaScript
+const validCID = 'QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF'
+
+const stream = ipfs.files.getReadableStream(validCID)
```
Comment (pgte) [inline SPEC/FILES.md]: shouldn't this be `ifs.lsPullStream` instead?

File: SPEC/FILES.md [hunk 7]
```diff
@@ -391,6 +391,158 @@ pull(
...
+```
+
+If no `callback` is passed, a promise is returned.
+
+**Example:**
+
+```JavaScript
+const validCID = 'QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF'
+
+ipfs.files.ls(validCID, function (err, files) {
```
Comment (pgte) [inline SPEC/FILES.md]: shouldn't this be `ipfs.ls` instead?

File: SPEC/FILES.md (hunk 6 above)
Comment (diasdavid) [reply SPEC/FILES.md]: thanks for the catch! :)

**Comment by mitra42** [comment]
Does this fix the bug "c" from https://github.com/ipfs/interface-ipfs-core/issues/164 which you commented was being worked on here?

**Comment by Beanow** [comment]
Would this be what you're looking for?

https://github.com/ipfs/interface-ipfs-core/blob/dd922523e32ae1d99ed63d1bd18d4e5cb7a20323/SPEC/FILES.md#get
```js
const validCID = 'QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF'

ipfs.files.get(validCID, function (err, files) {
  files.forEach((file) => {
    console.log(file.path)
    console.log(file.content.toString('utf8')) // <-- file.content is a Buffer
  })
})
```

**Comment by diasdavid** [comment]
@mitra42 yes :)

**Comment by diasdavid** [comment]
Next:
- https://github.com/ipfs/js-ipfs-api/pull/619
- https://github.com/ipfs/js-ipfs/pull/1086

**Comment by mitra42** [comment]
@Beanow That case you've given is another case, i.e. where the hash refers to multiple files and having .content available would be a much nicer interface than having to use streams. Does that work if it was a single file. My two tests are with QmTds3bVoiM9pzfNJX6vT2ohxnezKPdaGHLd4Ptc4ACMLa (a file that should be 184324 bytes and therefore not sharded - and failed previous files.cat.
Qmbzs7jhkBZuVixhnM3J3QhMrL6bcAoSYiRPZrdoX3DhzB single sharded file which works in files.cat
And I also test with a simple json file uploaded with dag.put, which i think would break your files case as it will throw an "invalid node type" error.

**Pull request merged**